Library/Homebrew/test/api/internal_tap_json/formula_spec.rb

```diff
@@ -0,0 +1,128 @@
+        ln_s(core_tap/"Formula/p/ponyc.rb", core_tap/"Aliases/ponyc-lang")
+      end
+    end
+
+    it "creates the expected hash" do
+      api_hash = CoreTap.instance.to_api_hash
+      api_hash["tap_git_head"] = tap_git_head # tricky to mock
+
+      expect(JSON.generate(api_hash)).to eq(internal_tap_json)
+    end
+  end
+
+  context "when loading JSON" do
+    before do
+      ENV["HOMEBREW_INTERNAL_JSON_V3"] = "1"
+      ENV.delete("HOMEBREW_NO_INSTALL_FROM_API")
+
+      allow(Homebrew::API).to receive(:fetch_json_api_file)
+        .with("internal/v3/homebrew-core.jws.json")
+        .and_return(JSON.parse(internal_tap_json), false)
+    end
+
+    it "loads tap aliases" do
+      expect(CoreTap.instance.alias_table).to eq({
+        "fennel-lang" => "fennel",
+        "ponyc-lang"  => "ponyc",
+      })
+    end
+
+    it "loads formula renames" do
+      expect(CoreTap.instance.formula_renames).to eq({
+        "advancemenu" => "advancemame",
+        "amtk"        => "libgedit-amtk",
+        "annie"       => "lux",
+        "antlr2"      => "antlr@2",
+        "romanesco"   => "fennel",
+      })
+    end
+
+    it "loads tap migrations" do
+      expect(CoreTap.instance.tap_migrations).to eq({
+        "adobe-air-sdk"          => "homebrew/cask",
+        "android-ndk"            => "homebrew/cask",
+        "android-platform-tools" => "homebrew/cask",
+        "android-sdk"            => "homebrew/cask",
+        "app-engine-go-32"       => "homebrew/cask/google-cloud-sdk",
+      })
+    end
+
+    it "loads tap git head" do
+      expect(Homebrew::API::Formula.tap_git_head)
+        .to eq(tap_git_head)
+    end
+
+    context "when loading formulae" do
+      let(:fennel_metadata) do
+        {
+          "dependencies"     => ["lua"],
+          "desc"             => "Lua Lisp Language",
+          "full_name"        => "fennel",
+          "homepage"         => "https://fennel-lang.org",
+          "installed"        => [],
+          "license"          => "MIT",
+          "name"             => "fennel",
+          "ruby_source_path" => "Formula/f/fennel.rb",
+          "tap"              => "homebrew/core",
+          "tap_git_head"     => tap_git_head,
+          "versions"         => { "bottle"=>true, "head"=>nil, "stable"=>"1.4.0" },
+        }
+      end
+
+      let(:ponyc_metadata) do
+        {
+          "desc"                   => "Object-oriented, actor-model, capabilities-secure programming language",
+          "full_name"              => "ponyc",
+          "homepage"               => "https://www.ponylang.io/",
+          "license"                => "BSD-2-Clause",
+          "name"                   => "ponyc",
+          "ruby_source_path"       => "Formula/p/ponyc.rb",
+          "tap"                    => "homebrew/core",
+          "tap_git_head"           => tap_git_head,
+          "uses_from_macos"        => [{ "llvm"=>[:build, :test] }, "zlib"],
+          "uses_from_macos_bounds" => [{}, {}],
```

Comment:
Sorry to increase up scope here potentially (and can be in another PR) but: this API is fairly awful due to backwards compatibility concerns we don't have for v3 so would be nice to address. CC @Bo98 for thoughts on what that should look like.

Reply:
Yeah we were going to cycle back and clean this up. I can open a PR with suggestions

Reply:
Cool: given that: fine to merge with this as-is in this PR 👍🏻. A TODO or something might be nice though?

Reply:
Can a TODO be added here please.

Reply:
I left a not in the original issue thread: https://github.com/Homebrew/brew/issues/16410#issuecomment-1963079892

Reply:
@apainintheneck I'd really love an inline Ruby comment here if possible.

Reply:
It seems weird to leave a comment when any potential changes are exploratory. We haven't decided on the form that we want yet.

Reply:
@apainintheneck It's a hard `TODO` because we're not going to ship the internal JSON v3 without changing this. I'm fine on not blocking the PR on this stuff but I would really like something beyond a comment in another issue to track this in the actual code.

Reply:
```suggestion
          # TODO: improve this API before we ship internal API v3 to users
          "uses_from_macos"        => [{ "llvm"=>[:build, :test] }, "zlib"],
          "uses_from_macos_bounds" => [{}, {}],
```

Reply:
> @apainintheneck It's a hard `TODO` because we're not going to ship the internal JSON v3 without changing this. I'm fine on not blocking the PR on this stuff but I would really like something beyond a comment in another issue to track this in the actual code.

I'm fine with the comment but this is news to me. I didn't know that this was that important either way.

Reply:
Apologies for my unclear communication! I'm trying to get the balance right between being overly pedantic/demotivating and blocking individual PRs vs. making clear what's required before we ship to users. Great work getting this shipped, thanks @apainintheneck ❤️